compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java

```diff
@@ -1652,7 +1652,8 @@ public void setForeachTypedBindingPatternType(BLangForeach foreachNode) {
                     break;
                 }
                 varType = streamType.constraint;
-                if (streamType.error != null) {
+                List<BType> completionType = getAllTypes(streamType.error);
+                if (completionType.stream().anyMatch(type -> type.tag != TypeTags.NIL)) {
```

Comment:
The spec says for the collection in a foreach statement

> The static type of action-or-expr must be an iterable type with an **iteration completion type of nil**.

Since `never` is a subtype of `()`, I think this should work with `never` too. WDYT? We can change this separately though.

Reply:
Will handle this separately after clarifying. https://github.com/ballerina-platform/ballerina-lang/issues/29758 will be used to track this